### PR TITLE
feat(a11y): improve chart figure

### LINKED ISSUE #1099: More accessible chart figure
## Describe the bug
Charts provide next-to-no information to screen readers. 

## Expected behavior
Provide a robust structure that can accommodate more chart info being added to it.

## Implementation details
Turn `<canvas role="img" />` into a more robust representation:
```jsx
<figure>
	<canvas role="presentation">
		<dl class="visually-hidden" {...ariaProps}>
			<dt>Chart type</dt>
			<dd>{chartType}</dd>
		</dl>
	</canvas>
	<Legend />
</figure>
```

- This structure may seem a little overkill but the ultimate goal is something in the vein of [this PoC](https://codepen.io/myasonik/pen/yLVXpKr?editors=0010) so it sets up a good foundation.
- Note, the `aria-*` props moved from the `<canvas />` to the `<dl />`
- I'd recommend tackling this after https://github.com/elastic/elastic-charts/issues/1096, https://github.com/elastic/elastic-charts/issues/1097, and https://github.com/elastic/elastic-charts/issues/1098 otherwise it's too much structure for too little info

### PR DESCRIPTION
**Pull request by rshen91**
## Summary
Closes #1099 

Adding series types for Cartesian charts.

### Checklist

Delete any items that are not applicable to this PR.
- [X] Unit tests were updated or added to match the most common scenarios
Unfortunately the previous test in the accessibility.test.tsx tests has been removed 
```
it('should include the series types if one type of series', async () => {
    const tree = await common.testAccessibilityTree(
      'http://localhost:9001/iframe.html?id=annotations-lines--x-continuous-domain',
      '.echCanvasRenderer',
    );
    // the legend has bars and lines as value.descriptions not value.name
    const hasTextOfChartTypes = tree.children.filter((value) => {
      return value.name === 'bars';
    });
    expect(hasTextOfChartTypes.length).toBe(1);
  });
```

The issue is that the accessibilitySnapshot is not capturing the aria-label consistently in the testAccessibilityTree function in common 

```
await page.accessibility.snapshot().then((value) => {
      return value;
    });
```

Luckily, we are capturing the aria-label in `chart.test.tsx.snap`

### CODE REVIEW

**Comment by rshen91** [comment]
jenkins test this please

File: integration/tests/accessibility.test.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { common } from '../page_objects/common';
+
+describe('Accessibility tree', () => {
```
Comment (nickofthyme) [inline integration/tests/accessibility.test.ts:22]: 😍

File: src/chart_types/xy_chart/renderer/dom/_screen_reader.scss [hunk 2]
```diff
@@ -0,0 +1,7 @@
+.screen-reader {
+  position: absolute;
+  left: -10000px;
+  top: auto;
+  width: 1px;
+  height: 1px;
```
Comment (nickofthyme) [inline src/chart_types/xy_chart/renderer/dom/_screen_reader.scss:6]: Does this need to be visible to work with the screen reader? we could add `pointer-events: none` and `visibility: hidden`?

```suggestion
  position: absolute;
  pointer-events: none;
  visibility: hidden;
  width: 1px;
  height: 1px;
```

File: src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx [hunk 3]
```diff
@@ -152,27 +152,50 @@ class XYChartComponent extends React.Component<XYChartProps> {
...
+          // eslint-disable-next-line jsx-a11y/no-interactive-element-to-noninteractive-role
+          role="presentation"
+        >
+          <dl className="screen-reader">
+            {seriesTypes.length > 0 ? (
+              <dt>
+                Chart type(s)
+                {seriesTypes.map((value, index) => {
+                  return <dd key={`${index}`}>{value}</dd>;
+                })}
```
Comment (nickofthyme) [inline src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx:191]: simplified...

```suggestion
                {seriesTypes.map((value, index) => <dd key={`${index}`}>{value}</dd>)}
```

**Review by nickofthyme — APPROVED** [review]
Overall, LGTM. I just left a few comments.

File: src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx [hunk 4]
```diff
@@ -152,27 +152,50 @@ class XYChartComponent extends React.Component<XYChartProps> {
...
+              </dt>
+            ) : (
+              <dt>No series in chart</dt>
+            )}
+          </dl>
+        </canvas>
+      </figure>
```
Comment (markov00) [inline src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx:198]: few words here:
- the `dl` tag represents a `description list` where we should include subsequent pairs of `dt` (description term) and `dd` (description details). https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl 
Having a `term` with no `details` looks like semantically wrong to me. and instead of having `<dt>No series in chart</dt>` we can also remove the `dl` completely if we don't provide more info. I'm also pretty sure that if we don't have series in chart, the chart is not rendered and we have a "no data available" text instead.
- the screen readers are not very smart and a text like `Chart type(s)` is probably read as `chart type open parentesis s closed parentesis`. As we know how many chart types we can be smarter and have something like.
```html
<dt>Chart type</dt>
<dd>Bar chart</dt>
```
or, when we have multiple types:
```html
<dt>Chart type</dt>
<dd>Mixed chart: bar and line chart</dd>
```

**Review by markov00 — CHANGES_REQUESTED** [review]
Despite the fact that I like the approach taken, I think we should follow what was suggested in the original issue first.
`I'd recommend tackling this after #1096, #1097, and #1098 otherwise it's too much structure for too little info`

File: src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx [hunk 5]
```diff
@@ -152,27 +152,50 @@ class XYChartComponent extends React.Component<XYChartProps> {
...
       this.ctx = null;
       return null;
     }
 
+    const seriesTypes: string[] = [];
+    Object.entries(geometries).forEach((value) => {
+      if (value[1].length > 0) {
+        seriesTypes.push(value[0]);
+      }
+    });
```
[the remark this answers is not in the document]
Comment (myasonik) [reply src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx:168]: How would you recommend getting this data if not using the object keys?

On the i18n front, I don't think it's fair to expect a PR that improves a11y to  also solve Charts' i18n in one fell swoop. Without an existing i18n solution in place (correct me if I'm wrong as I'm only loosely familiar with the codebase), building with a hypothetical solution in mind brings a lot of overhead with little gain imo.

File: src/chart_types/xy_chart/renderer/dom/_screen_reader.scss (hunk 2 above)
Comment (myasonik) [reply src/chart_types/xy_chart/renderer/dom/_screen_reader.scss:6]: Hiding stuff has historically been fraught with bugs so it's best avoided at the time being. 

This almost matches [EUI's screen reader only mixin](https://github.com/elastic/eui/blob/master/src/global_styling/mixins/_helpers.scss#L90-L97) which has been battle tested for a while now.

The only thing that EUI has that this is missing is `overflow: hidden;` which I'd recommend adding to clip the text off from the 1px by 1px box we're shoving it into.

File: src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx (hunk 4 above)
Comment (myasonik) [reply src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx:198]: **RE: DL semantics**
Having a term with no details _is_ allowed but, I agree, it _is_ weird... (It also has some weird fall-through rules if there is a definition later...) 

If the chart isn't rendered with no data then I guess we don't need it anyways.

**RE: Chart types(s)**
By default, most screen readers actually ignore punctuation and communicate much of it by intonation or specific tones. But generally we shouldn't worry about how a specific screen reader is going to read something, because, however it does it, it is ultimately expected by the user. We don't need to try to preempt a user's assistive tech with our own assumptions. 

That said, I do like your adjusted phrasing though. 😄 It's more definitive and flows a little nicer.

**Comment by myasonik** [comment]
> Despite the fact that I like the approach taken, I think we should follow what was suggested in the original issue first.
> > I'd recommend tackling this after #1096, #1097, and #1098 otherwise it's too much structure for too little info

@markov00 I know I wrote that ☝️ but this PR still does improve the state of charts a11y so I'd rather merge it (when it's green) and do the issues in a different order than have to sit on this PR while it rots.

It does actually mean less work in the long run imo because this PR changes the DOM structure somewhat and the other work can be built directly into this new structure instead of needing to refactor it after the fact.

**TL;DR** The new structure is arguably a little overkill... But that's better than the vast gap we have now...

File: src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx (hunk 4 above)
Comment (rshen91) [reply src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx:197]: Please see e7d1230 - thank you!

File: src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx [hunk 6]
```diff
@@ -152,27 +152,50 @@ class XYChartComponent extends React.Component<XYChartProps> {
...
+        <canvas
+          ref={forwardStageRef}
+          className="echCanvasRenderer"
+          width={width * this.devicePixelRatio}
+          height={height * this.devicePixelRatio}
+          style={{
+            width,
+            height,
+          }}
+          aria-label="Chart"
```
[the remark this answers is not in the document]
Comment (rshen91) [reply src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx]: Please see e7d1230 thanks!

File: src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (rshen91) [reply src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx]: Refactored in e7d1230 thank you!

File: src/chart_types/xy_chart/renderer/dom/_screen_reader.scss [hunk 7]
```diff
@@ -0,0 +1,7 @@
+.screen-reader {
```
[the remark this answers is not in the document]
Comment (rshen91) [reply src/chart_types/xy_chart/renderer/dom/_screen_reader.scss]: good call thank you - fixed in e7d1230

File: src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx (hunk 3 above)
Comment (rshen91) [reply src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx]: thank you fixed in e7d1230

File: src/chart_types/xy_chart/renderer/dom/_screen_reader.scss (hunk 2 above)
Comment (rshen91) [reply src/chart_types/xy_chart/renderer/dom/_screen_reader.scss:6]: Thank you - please see e7d1230

File: src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (rshen91) [reply src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx]: In e7d1230 I ended up removing the aria-label because it's not read by VoiceOver and this PR adds more information. I was leaning towards not having the aria-label since it's not providing more information on the `<dl>` tags.

File: src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx]: > How would you recommend getting this data if not using the object keys?

I'm referring to the fact that we are extracting the chart type "strings" from an object `geometries` that is an internal object which structure can change in the future easily. 

This refactoring increase also the current issue: if the `geometries` object change its signature, the `getNameFunction` can just return `unkown chart` for each chart type without being warned.

```
Object.entries(geometries).forEach((value) => {
      if (value[1].length > 0 && value[0]) {
        seriesTypes.push(getNameFunction(value[0]));
      }
    });
....
export const getNameFunction = (key: string): string => {
  const screenReader: Map<string, string> = new Map();
  screenReader
    .set('bars', 'Bar chart')
    .set('areas', 'Area chart')
    .set('lines', 'Line chart')
    .set('bubbles', 'Bubble chart');
  return screenReader.get(key) ?? 'unknown chart';
};
```

The concept of chart type is already built into each `spec`, the react component that specifies the chart configuration. I'm saying that. instead of taking the chart type information from an internal object, like `geometris`, we should create that information from the actual specs using an already well-known and exposed enum called `SeriesType`.

A selector that picks the actual series types and creates a Set of unique `SeriesType`s is cleaner and we can map the well-known `SeriesType` enum values to a set of internationalized strings.

[the remark this answers is not in the document]
Comment (rshen91) [reply src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx]: Ok thanks please see 9e036d3

**Comment by rshen91** [comment]
jenkins test this please

**Comment by rshen91** [comment]
jenkins test this please

**Review by myasonik — APPROVED** [review]
🚀

**Review by markov00 — APPROVED** [review]
Looks good to me, I've added few minor comments

File: src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx [hunk 8]
```diff
@@ -152,27 +154,38 @@ class XYChartComponent extends React.Component<XYChartProps> {
...
       chartContainerDimensions: { width, height },
+      seriesTypes,
     } = this.props;
 
     if (!initialized || isChartEmpty) {
       this.ctx = null;
       return null;
     }
+    const multipleSeriesTypes: string[] = [];
+    seriesTypes.forEach((value) => multipleSeriesTypes.push(value));
```
[the remark this answers is not in the document]
Comment (rshen91) [reply src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx]: Thank you! 9b11e85

File: src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx [hunk 9]
```diff
@@ -152,27 +154,38 @@ class XYChartComponent extends React.Component<XYChartProps> {
...
+          height={height * this.devicePixelRatio}
+          style={{
+            width,
+            height,
+          }}
+          // eslint-disable-next-line jsx-a11y/no-interactive-element-to-noninteractive-role
+          role="presentation"
+        >
+          <dl className="echScreen-reader">
+            <dt> Chart type </dt>
```
[the remark this answers is not in the document]
Comment (rshen91) [reply src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx]: Thank you! 9b11e85

File: src/chart_types/xy_chart/renderer/dom/_screen_reader.scss [hunk 10]
```diff
@@ -0,0 +1,8 @@
+.echScreen-reader {
```
[the remark this answers is not in the document]
Comment (rshen91) [reply src/chart_types/xy_chart/renderer/dom/_screen_reader.scss]: Good call - 9b11e85

**Comment by rshen91** [comment]
Jenkins, test this please
